File: packages/qiskit-bridge/algorithms/shor.py
"""Shor de referencia con Qiskit: QPE de order-finding.

La multiplicacion modular se aporta como UnitaryGate (validacion independiente
del motor C++). La QFT inversa se implementa a mano con la MISMA convencion que
el motor C++ para permitir comparacion elemento a elemento del registro de
conteo. Layout: work [0, nw), counting [nw, nw+t).
"""

from __future__ import annotations

import math

import numpy as np
from qiskit import QuantumCircuit, transpile
from qiskit.circuit.library import UnitaryGate
from qiskit.quantum_info import Statevector
from qiskit_aer import AerSimulator


def _gcd(x: int, y: int) -> int:
    while y:
        x, y = y, x % y
    return x


def counting_qubits(N: int) -> int:
    n = math.ceil(math.log2(N))
    return 2 * n


def classical_order(a: int, N: int) -> int:
    if _gcd(a, N) != 1:
        return 0
    r, cur = 1, a % N
    while cur != 1:
        cur = (cur * a) % N
        r += 1
        if r > N:
            return 0
    return r


def _modmul_unitary(mult: int, N: int, nw: int) -> np.ndarray:
    """Matriz de permutacion de |y> -> |mult*y mod N> (identidad si y>=N)."""
    dim = 1 << nw
    P = np.zeros((dim, dim), dtype=complex)
    for y in range(dim):
        y2 = (mult * y) % N if y < N else y
        P[y2, y] = 1.0
    return P


def _inverse_qft(qc: QuantumCircuit, start: int, t: int) -> None:
    """QFT inversa con la misma convencion que el motor C++ (qubit start = LSB)."""
    for i in range(t // 2):
        qc.swap(start + i, start + t - 1 - i)
    for j in range(t):
        for k in range(j):
            qc.cp(-math.pi / (2 ** (j - k)), start + k, start + j)
        qc.h(start + j)


def build_circuit(N: int, a: int) -> tuple[QuantumCircuit, int, int]:
    nw = math.ceil(math.log2(N))
    t = counting_qubits(N)
    qc = QuantumCircuit(nw + t)
    work = list(range(nw))
    count_start = nw

    qc.x(work[0])  # registro de trabajo = |1>
    for j in range(t):
        qc.h(count_start + j)  # superposicion en el conteo

    for j in range(t):
        mult = pow(a, 1 << j, N)
        gate = UnitaryGate(_modmul_unitary(mult, N, nw), label=f"*{mult}%{N}").control(1)
        qc.append(gate, [count_start + j] + work)

    _inverse_qft(qc, count_start, t)
    return qc, nw, t


def run_shor(N: int, a: int) -> dict:
    """Ejecuta la QPE de Shor y devuelve la distribucion del registro de conteo.

    Retorna:
        {
            "counting_probabilities": [float, ...],  # 2^t valores
            "counting_qubits": int,
            "work_qubits": int,
            "order": int,          # periodo clasico (referencia)
            "factors": [int, ...]  # factores clasicos (referencia)
        }
    """
    if _gcd(a, N) != 1:
        g = _gcd(a, N)
        return {
            "counting_probabilities": [],
            "counting_qubits": 0,
            "work_qubits": 0,
            "order": 0,
            "factors": sorted({g, N // g}),
        }

    qc, nw, t = build_circuit(N, a)
    sim = AerSimulator(method="statevector")
    qc.save_statevector()
    sv = Statevector(sim.run(transpile(qc, sim)).result().get_statevector().data)

    # Statevector.probabilities(qargs) toma qargs[0] como bit mas significativo;
    # el motor C++ usa el primer qubit de conteo como bit menos significativo, asi
    # que invertimos el orden para que el marginal coincida elemento a elemento.
    counting = list(range(nw + t - 1, nw - 1, -1))
    probs = sv.probabilities(counting)  # marginal del registro de conteo

    r = classical_order(a, N)
    factors: list[int] = []
    if r and r % 2 == 0:
        x = pow(a, r // 2, N)
        if x != N - 1:
            for f in (_gcd(x + 1, N), _gcd(x - 1, N)):
                if f not in (1, N):
                    factors.append(f)
    return {
        "counting_probabilities": [float(p) for p in probs],
        "counting_qubits": t,
        "work_qubits": nw,
        "order": r,
        "factors": sorted(set(factors)),
    }

File: packages/core-engine/src/shor.cpp
#include "shor.h"
#include <cmath>
#include <algorithm>
#include <stdexcept>

namespace shor {

// Multiplicacion modular portable: usa enteros de 128 bits en GCC/Clang y cae a
// 64 bits en MSVC (que no tiene __uint128_t). Exacta mientras a,b < 2^32, lo que
// se cumple de sobra para las N pequenas de esta app.
static inline uint64_t mulmod(uint64_t a, uint64_t b, uint64_t m) {
#if defined(__SIZEOF_INT128__)
    return static_cast<uint64_t>((unsigned __int128)a * b % m);
#else
    return a * b % m;
#endif
}

uint64_t gcd(uint64_t x, uint64_t y) {
    while (y) {
        uint64_t t = x % y;
        x = y;
        y = t;
    }
    return x;
}

uint64_t powmod(uint64_t base, uint64_t exp, uint64_t mod) {
    uint64_t result = 1 % mod;
    base %= mod;
    while (exp) {
        if (exp & 1) result = mulmod(result, base, mod);
        base = mulmod(base, base, mod);
        exp >>= 1;
    }
    return result;
}

uint64_t classicalOrder(uint64_t a, uint64_t N) {
    if (gcd(a, N) != 1) return 0;
    uint64_t r = 1, cur = a % N;
    while (cur != 1) {
        cur = mulmod(cur, a, N);
        ++r;
        if (r > N) return 0;  // salvaguarda
    }
    return r;
}

int countingQubits(uint64_t N) {
    int n = 0;
    while ((uint64_t(1) << n) < N) ++n;  // ceil(log2 N)
    return 2 * n;
}

// Inversa de la QFT sobre el registro de conteo [start, start+t).
// Convencion: qubit `start` es el bit menos significativo del entero medido.
static void inverseQFT(QuantumStateVector& sv, int start, int t) {
    // Invertir el orden de bits.
    for (int i = 0; i < t / 2; ++i) sv.applySwap(start + i, start + t - 1 - i);
    for (int j = 0; j < t; ++j) {
        for (int k = 0; k < j; ++k) {
            const double angle = -M_PI / std::pow(2.0, j - k);
            sv.applyCPhase(start + k, start + j, angle);
        }
        sv.applyHadamard(start + j);
    }
}

// Fracciones continuas: mejor r <= N con |m/M - s/r| pequeno. Devuelve el
// denominador de la convergente cuyo r verifica a^r = 1 mod N, o 0.
static uint64_t orderFromPhase(uint64_t m, uint64_t M, uint64_t a, uint64_t N) {
    if (m == 0) return 0;
    // Expansion en fraccion continua de m/M.
    uint64_t num = m, den = M;
    uint64_t h_prev = 0, h_cur = 1;  // numeradores convergentes
    uint64_t k_prev = 1, k_cur = 0;  // denominadores convergentes
    while (den != 0) {
        uint64_t q = num / den;
        uint64_t h_next = q * h_cur + h_prev;
        uint64_t k_next = q * k_cur + k_prev;
        // k_next es un candidato a r.
        if (k_next > 0 && k_next <= N && powmod(a, k_next, N) == 1) return k_next;
        h_prev = h_cur;
        h_cur = h_next;
        k_prev = k_cur;
        k_cur = k_next;
        uint64_t r = num % den;
        num = den;
        den = r;
        if (k_cur > N) break;
    }
    return 0;
}

Result run(uint64_t N, uint64_t a, const FrameCallback& on_frame) {
    Result res;
    res.N = N;
    res.a = a;
    if (N < 3 || a < 2 || a >= N) throw std::invalid_argument("Shor: 2 <= a < N, N >= 3");

    // Caso trivial: a comparte factor con N.
    uint64_t g = gcd(a, N);
    if (g != 1) {
        res.order = 0;
        res.factors = {g, N / g};
        res.success = true;
        on_frame(0, true, Eigen::VectorXd::Zero(1));
        return res;
    }

    const int nw = static_cast<int>(countingQubits(N) / 2);  // ceil(log2 N)
    const int t = countingQubits(N);                         // registro de conteo
    res.counting_qubits = t;
    res.work_qubits = nw;

    if (t + nw > 20) throw std::invalid_argument("Shor: N demasiado grande (>20 qubits)");

    // Layout: work [0, nw), counting [nw, nw+t).
    const int work_start = 0;
    const int count_start = nw;
    QuantumStateVector sv(t + nw);

    // Registro de trabajo inicializado a |1>.
    sv.applyX(work_start);
    // Superposicion en el registro de conteo.
    for (int j = 0; j < t; ++j) sv.applyHadamard(count_start + j);
    on_frame(0, false, sv.marginalProbabilities(count_start, t));

    // QPE: controlled-U^(2^j), con U|y> = |a*y mod N>.
    for (int j = 0; j < t; ++j) {
        const uint64_t mult = powmod(a, uint64_t(1) << j, N);  // a^(2^j) mod N
        sv.applyControlledModMul(count_start + j, work_start, nw, mult, N);
    }
    // QFT inversa sobre el registro de conteo.
    inverseQFT(sv, count_start, t);

    const Eigen::VectorXd counting = sv.marginalProbabilities(count_start, t);
    on_frame(1, true, counting);

    // Post-proceso: examinar los picos mas probables.
    const uint64_t M = uint64_t(1) << t;
    std::vector<int> idx(counting.size());
    for (int i = 0; i < counting.size(); ++i) idx[i] = i;
    std::sort(idx.begin(), idx.end(), [&](int p, int q) { return counting(p) > counting(q); });

    uint64_t r = 0;
    for (int rank = 0; rank < std::min<int>(8, idx.size()); ++rank) {
        uint64_t cand = orderFromPhase(idx[rank], M, a, N);
        if (cand) {
            r = cand;
            break;
        }
    }
    if (r == 0) {
        res.success = false;
        return res;
    }
    res.order = r;

    // Factores a partir del periodo: r par y a^{r/2} != -1 mod N.
    if (r % 2 == 0) {
        uint64_t x = powmod(a, r / 2, N);
        if (x != N - 1) {
            uint64_t f1 = gcd(x + 1, N);
            uint64_t f2 = gcd(x + N - 1, N);
            std::vector<uint64_t> fs;
            for (uint64_t f : {f1, f2})
                if (f != 1 && f != N) fs.push_back(f);
            if (!fs.empty()) {
                std::sort(fs.begin(), fs.end());
                fs.erase(std::unique(fs.begin(), fs.end()), fs.end());
                res.factors = fs;
                res.success = true;
            }
        }
    }
    return res;
}

}  // namespace shor
